### ISSUE TITLE
#define RHEL_RELEASE_CODE 0 affects compatiblity

### ISSUE THREAD

**Issue by konradknitter** [vmware/photon#1364]
### Describe the bug

The define:

+#ifndef RHEL_RELEASE_CODE
+#define RHEL_RELEASE_CODE 0
+#endif

introduced in 
https://github.com/vmware/photon/blob/master/SPECS/linux/amzn/0065-Add-Amazon-EFA-driver-version-1.4.patch
breaks compatibility for components that have conditions based on particular RHEL versions.

Patches as:
https://github.com/vmware/photon/blob/master/SPECS/linux/ice/0001-ice-Use-PTP_SYS_OFFSET_EXTENDED_IOCTL-support.patch
https://github.com/vmware/photon/blob/master/SPECS/linux/iavf/0001-iavf-Use-PTP_SYS_OFFSET_EXTENDED_IOCTL-support.patch
as well as part of https://github.com/vmware/photon/blob/1826d1a781fcae1ae11c634748c3d4253175ddd3/SPECS/linux/i40e/0001-Add-support-for-gettimex64-interface.patch
would not be needed if above would be resolved.

All of those drivers handle PTP correctly if RHEL_RELEASE_CODE is not defined (or have correct RHEL version)
There are many more RHEL_RELEASE_CODE checks that might (now or in the future) result in broken functionality, in those or other components.

### Reproduction steps

N/A

### Expected behavior

Isolate change required to use Amazon EFA driver to it's patch.

### Additional context

ICE, IAVF Patch would not be required. Part of patch i40e would not be required.

**Comment by sshedi** [comment]
Hi @konradknitter, what is the point of defining `RHEL_RELEASE_CODE` macro in Photon?
I say this is user error.

cc: @YustasSwamp @srivatsabhat

**Comment by srivatsabhat** [comment]
Hi @konradknitter 

Thank you for reaching out and for sharing your suggestions!


The definition of `RHEL_RELEASE_CODE` is part of the ena driver's `kcompat.h` and the macro is initialized to an invalid value (i.e., such as not to conflict with a RHEL-distro), so as to make the kcompat code work well.

For example, they intentionally define it to be `0` (if not already defined, like when building on a RHEL distro), so that they can use it for non-RHEL-specific compatibility code, such as this:

```
+#if !(RHEL_RELEASE_CODE)
+static inline u32 ethtool_rxfh_indir_default(u32 index, u32 n_rx_rings)
+{
+       return index % n_rx_rings;
+}
+#endif
```

So, I believe the code can be left as it is even on Photon OS, as the macro definition won't end up incorrectly signaling that the distro is RHEL, as explained above. In fact, the kcompat.h header has similar (intentionally invalid) initializations for other distro-specific macros too, such as Ubuntu and SUSE:

```
+#ifndef UTS_UBUNTU_RELEASE_ABI
+#define UTS_UBUNTU_RELEASE_ABI 0
+#define UBUNTU_VERSION_CODE 0

+#ifndef SLE_VERSION_CODE
+#define SLE_VERSION_CODE 0
+#endif /* SLE_VERSION_CODE */
```


In other words, this is how compatibility across multiple distros is handled in the ena driver -- this code is not supposed to be deleted by distros; instead they are expected to define any macros that are specific to their distro (but Photon has no such macros as of today; so, no changes are needed).


Further, this patch is only applied to the `linux-aws` flavor, whereas drivers such as `i40e`, `iavf` and `ice` are only built for other kernel flavors such as `linux`, `linux-esx` and `linux-rt` in Photon OS.



Let's look at the `ice` driver in more detail (similar analysis applies to `iavf` and `i40e` as well).


The code that is problematic to Photon OS is shown below:

src/kcompat.h in ice-1.8.3:


```
#if (RHEL_RELEASE_CODE && \
     (RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(7,7) && \
      RHEL_RELEASE_CODE < RHEL_RELEASE_VERSION(8,0)) || \
     (RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(8,1)))
#define HAVE_PTP_SYS_OFFSET_EXTENDED_IOCTL
#define HAVE_PTP_CLOCK_INFO_GETTIMEX64
#else /* RHEL >= 7.7 && RHEL < 8.0 || RHEL >= 8.1 */
struct ptp_system_timestamp {
        struct timespec64 pre_ts;
        struct timespec64 post_ts;
};

static inline void
ptp_read_system_prets(struct ptp_system_timestamp __always_unused *sts)
{
        ;
}

static inline void
ptp_read_system_postts(struct ptp_system_timestamp __always_unused *sts)
{
        ;
}
#endif /* !(RHEL >= 7.7 && RHEL != 8.0) */

```

Due to the way the above code is structured, any distro that is not RHEL and doesn't match the RHEL-versions in the check automatically gets the stub definitions of `ptp_read_system_prets()` etc., because it assumes that `PTP_SYS_OFFSET_EXTENDED_IOCTL` is not supported on that distro.

For upstream kernels, the following code takes effect, which marks `PTP_SYS_OFFSET_EXTENDED_IOCTL` and `PTP_CLOCK_INFO_GETTIMEX64` features as available only if the kernel version is >= v5.0

```
#else /* >= 5.0.0 */
#define HAVE_PTP_SYS_OFFSET_EXTENDED_IOCTL
#define HAVE_PTP_CLOCK_INFO_GETTIMEX64
...
#endif /* 5.0.0 */
```


However, just like RHEL, in Photon OS, we have backported the `PTP_SYS_OFFSET_EXTENDED_IOCTL` and `PTP_CLOCK_INFO_GETTIMEX64` features to Photon OS 3.0 which is based on the v4.19 kernel version. Therefore, none of these compatibility checks in the ice driver work properly on Photon OS -- they either add stub definitions (instead of making use of the full-fledged functionality already provided by Photon OS), or incorrectly assume that the features are not supported.


Therefore, to compile this driver **_correctly_** on Photon OS (i.e., making full use of `PTP_SYS_OFFSET_EXTENDED_IOCTL` and `PTP_CLOCK_INFO_GETTIMEX64` features backported to Photon OS v4.19 kernels), we need to export Photon OS-specific macros (just like RHEL, Ubuntu and SUSE) and extend `kcompat.h` in the `ice` driver to define `HAVE_PTP_SYS_OFFSET_EXTENDED_IOCTL` and `HAVE_PTP_CLOCK_INFO_GETTIMEX64` appropriately for Photon OS.

Regards,
Srivatsa

**Comment by konradknitter** [comment]
Hi,

thanks for the insides! 
It's clear for me now that I've had worked on incomplete information.
